### PR TITLE
Refactor and Cleanup Intellisense Implementation

### LINKED ISSUE #61: [Feature Request]  Add auto-completion popups (IntelliSense) for references in .bib files when typing \cite{...}
- An equivalent of this feature is available in Overleaf Webpage!

- **Would it be possible to add an auto-completion popup when using the command \cite{...}, loading references from .bib files contained in the project?**

Thanks!

### PR DESCRIPTION
**Pull request by iamhyc**
> resolves #61 
@QianrenLi you are suggested to participate in the whole refactor process.

### CODE REVIEW

**Review by QianrenLi — APPROVED** [review]

**Comment by QianrenLi** [comment]
OK and If any problem, I can do whatever I can to fix it.

**Review by QianrenLi — APPROVED** [review]

**Review by QianrenLi — APPROVED** [review]

**Review by QianrenLi — APPROVED** [review]

**Review by QianrenLi — APPROVED** [review]

**Review by QianrenLi — DISMISSED** [review]

**Review by QianrenLi — DISMISSED** [review]

**Review by QianrenLi — DISMISSED** [review]

**Review by QianrenLi — DISMISSED** [review]

**Review by QianrenLi — DISMISSED** [review]

**Review by QianrenLi — DISMISSED** [review]

File: src/intellisense/langCompletionProvider.ts [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+                return items;
+            default:
+                return [];
+        }
+    }
+
+    private async getReferenceCompletionItemsFromBib(vfs: VirtualFileSystem): Promise<vscode.CompletionItem[]> {
+        const bibList:string[] = this.symbolProvider.getBibList() ?? [];
+        const bibPaths = bibList.flatMap(
+            path => (path.split(',') ?? [])
```
Comment (iamhyc) [inline src/intellisense/langCompletionProvider.ts:406]: @QianrenLi why the path is joined with `,`. I cannot find the hint anywhere.

Comment (QianrenLi) [reply src/intellisense/langCompletionProvider.ts:406]: Sometimes, the path included in the macro `\bibgraphy` can be multiple file like `{ieee,reference}'.  The split is used to sperate them up.

**Comment by iamhyc** [comment]
check the functionality and then approve.

**Comment by iamhyc** [comment]
Ok, I will stop here today. You can check the correctness of the functionality now.

The whole logic is still a mess. I will continue clean it up after you confirm the above changes.

**Comment by iamhyc** [comment]
@QianrenLi Can you do it today?

**Review by QianrenLi — DISMISSED** [review]

**Comment by iamhyc** [comment]
@QianrenLi Can you check the correctness today?

**Comment by iamhyc** [comment]
@QianrenLi Can you check the correctness today?

**Comment by iamhyc** [comment]
@QianrenLi Please remove the approval status.

**Comment by iamhyc** [comment]
@QianrenLi Now, could you please check the correctness of the functionality today?

**Comment by QianrenLi** [comment]
Seems to have some issue with the class extend, I will have it a look today

**Comment by iamhyc** [comment]
> Seems to have some issue with the class extend, I will have it a look today

You do not check the code implementation. You only check the functionality correctness.

**Comment by iamhyc** [comment]
> @QianrenLi Now, could you please check the correctness of the functionality today?

@QianrenLi So, can you check the *functionality correctness* today? Give me a firm answer.

**Comment by QianrenLi** [comment]
The logic in the symbol provider seems good to me, as for the reset part, I'll take it a look.

**Comment by iamhyc** [comment]
> > @QianrenLi Now, could you please check the correctness of the functionality today?
> 
> @QianrenLi So, can you check the _functionality correctness_ today? Give me a firm answer.

@QianrenLi So, can you check the ***functionality correctness*** today? Give me a firm answer.

You have no right to comment on code implementation now.

**Review by QianrenLi — DISMISSED** [review]

**Comment by iamhyc** [comment]
If you are interested, I have not started to refactor the whole logic implementation.

**Comment by iamhyc** [comment]
@QianrenLi Dare you approve those changes before confirming the functionality correctness!

**Comment by QianrenLi** [comment]
I am not able to read the logic in other part since I am not familiar with it, as for the symbol provider, as mentioned, I believe is good enough.

**Comment by iamhyc** [comment]
> I am not able to read the logic in other part since I am not familiar with it, as for the symbol provider, as mentioned, I believe is good enough.

CHECK THE CORRECTNESS, NO CODE REVIEW.

DO YOU COPY? CANT YOU READ?

**Comment by iamhyc** [comment]
In case you really cannot read, I will do the translation for you!

> Now, could you please check the correctness of the functionality today?

现在，您能检查一下功能的正确性吗？

> You do not check the code implementation. You only check the functionality correctness.

您不检查代码执行。只检查功能的正确性。

> So, can you check the functionality correctness today? Give me a firm answer.
> You have no right to comment on code implementation now.

那么，你今天能检查功能的正确性吗？给我一个明确的答复。

你现在无权评论代码的实现。

> If you are interested, I have not started to refactor the whole logic implementation.
如果你感兴趣，我还没有开始重构整个逻辑实现。

> CHECK THE CORRECTNESS, NO CODE REVIEW.
> DO YOU COPY? CANT YOU READ?

检查正确性，无代码审查。

你理解了吗？你看不懂吗？

**Comment by iamhyc** [comment]
@QianrenLi Let me *try* to make it more clear:

RUN THE EXTENSION, DO THE TEST.

**Comment by iamhyc** [comment]
@QianrenLi if you are still not clear about you role in this refactor, keep asking.

I need to know if you can do the test today.

**Comment by iamhyc** [comment]
@QianrenLi Give me your final answer before 14:00 UTC+8 today.

**Comment by QianrenLi** [comment]
> Seems to have some issue with the class extend, I will have it a look today

If the correctness means runable, the code is not runnable which it shows things like the extend undefined class in the vscode.

If the correctness means the symbol provider is logically correct if the system is able to run, I believe it might be correct. At this stage, I'm still looking to how to get the code run which is the basis of the further checking on the actual performance.

**Comment by iamhyc** [comment]
@QianrenLi yes or no? can or cannot?

**Comment by iamhyc** [comment]
I am only asking one simple question, CAN you answer it? @QianrenLi

**Comment by QianrenLi** [comment]
OK, the code is not runnable, thus I cannot test it.

**Comment by iamhyc** [comment]
> OK, the code is not runnable, thus I cannot test it.

then it sbould be: yoi test it, and it is wrong.

now tell me the casue and talk nothing about your own opinion.

**Comment by QianrenLi** [comment]
Current code shows 'Activating extension 'iamhyc.overleaf-workshop' failed: Class extends value undefined is not a constructor or null.' in terminal, which seems to happen in the constructor in `index.ts`. 
Any provider created here 
image
will lead to this problem. If no provider is created, the problem is not happen.

**Comment by iamhyc** [comment]
> Current code shows 'Activating extension 'iamhyc.overleaf-workshop' failed: Class extends value undefined is not a constructor or null.' in terminal, which seems to happen in the constructor in `index.ts`. Any provider created here image will lead to this problem. If no provider is created, the problem is not happen.

ok, good, i will fix it.

**Comment by iamhyc** [comment]
@QianrenLi Try the latest code now.

**Comment by QianrenLi** [comment]
Both symbol provider and bib item completion works normally.

**Comment by iamhyc** [comment]
@QianrenLi Glad that you talk nothing about your opinion this time.

Now question: Can you please check the *functionality correctness* now? The reply should fall into the following three situation:

1) Yes, I can. There is nothing wrong with the functionality after thorough test I can try.

2) Yes, I can. There is something wrong with the functionality. The details are listed below
- (error 1)
- (error 2)
- ...

3) No, I can't. I am doing something else.

**Comment by QianrenLi** [comment]
Yes, I can. There is nothing wrong with the functionality after thorough test I can try.

**Comment by iamhyc** [comment]
> Yes, I can. There is nothing wrong with the functionality after thorough test I can try.

ok, good, then I will start to refactor the logic part. You can unsubscribe now. I will let you know when a new test is needed.

If you have time, you can try to learn something from the refactor process, by skimming through the commit list.

**Comment by iamhyc** [comment]
@QianrenLi I have done the refactor. Please check the functionality correctness (by run and text the packaged extension).

If you have time, you can also do the code review.

If there is no problem, you can click the approve button. Otherwise, please let me know and I will fix it.

**Review by QianrenLi — APPROVED** [review]
Function good and no problem.

**Comment by iamhyc** [comment]
> Function good and no problem.

ok, thank you.

**Pull request merged**